### PR TITLE
Fix style constraint save when strength is NotSet

### LINKED ISSUE #58341: Georeferencer Dialog Not Working with 3.34.8 LTR
### What is the bug or the crash?

The enter coordinates window does not open properly in 3.34.8 - please note this is not on all computers, two members of my team have seen the issue, the other two but (as far as I am am aware) we all have the same set up.
georeferencer

### Steps to reproduce the issue

adding control points when georeferencing - consistent every time.

### Versions

3.34.8 only (we have reverted to 3.28 and it works fine in that)

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

we always only download the LTR.

### LINKED ISSUE #58431: Unwanted unique constraint added when exporting to geopackage, if not null constraint is applied via default style.
### What is the bug or the crash?

When a layer with a default style is exported to a geodatabase, the "Unique" and "Enforce unique constraint" constraints get turned on in the form widget for all columns that have the "Not null" constraint applied in the default style. 

### Steps to reproduce the issue

1) Open a map layer, from a PostGIS geodatabase schema, in QGIS.
2) Open the Layer Properties wizard for your layer, and go to the Attributes Form tab. Set the dropdown at the top to "Drag and Drop Designer".
3) In the Available Widgets box, select one of the columns in your layer to open the widget settings.
4) In the Constraints tab, tick Not null. **Leave Unique unticked.**
5) Click "Apply".
5) Click the "Style" tab at the bottom of the window > "Save as Default" > Datasource Database.
6) Close the wizard window.
6) Remove the layer from your project. Then re-add it. The default style should load,  and the constraints in the widgets will be set up correctly.
7) Open the Package layers tool from the Processing Toolbox.
8) Export your layer to a geopackage (choose a folder location), with "save layer styles into Geopackage" ticked
9) Open your new geopackage in QGIS, and check the constraints on the columns in the Attributes Form in the Layer Properties. **You will find that, for any columns that had "Not null" constraint ticked, but not the unique constraint, they now have the "Unique" and "Enforce unique" constraints ticked as well.**
....
Note that if, after opening your map layer with the default style, you then reload the exact same style using the Style tab at the bottom of the Layer Properties window (Style > Load Style... > From Database (postgres) > choose your default style in the "Styles related to the layer" box) then exporting it to the geopackage works perfectly! Note also that in the example in the screenshot, the Not null comes from the database and is not enforced in the widget, but the same thing happens if it is only in the widget and if it is enforced.

NotNull_AndUniqe_Widget_Geopackage_layer
NotNull_NotUnique_Widget_PostGIS_layer
...
This is the log from the export to geopackage:
QGIS version: 3.16.5-Hannover
QGIS code revision: 58ba7c1ed6
Qt version: 5.11.2
GDAL version: 3.1.4
GEOS version: 3.8.1-CAPI-1.13.3
PROJ version: Rel. 6.3.2, May 1st, 2020
Processing algorithm…
Algorithm 'Package layers' starting…
Input parameters:
{ 'LAYERS' : ['dbname=\'bb_dev\' host=192.168.0.118 port=5432 sslmode=disable key=\'gid\' srid=27700 type=Point checkPrimaryKeyUnicity=\'1\' table=\"temp\".\"test_unique\" (geom)'], 'OUTPUT' : '//tsclient/P/LiveData/M1.1.19_DATA_MANAGEMENT/Temp_TestMergin/Test_unique_constraints.gpkg', 'OVERWRITE' : False, 'SAVE_STYLES' : True }

Packaging layer 1/1: test_unique
Execution completed in 2.01 seconds
Results:
{'OUTPUT': '//tsclient/P/LiveData/M1.1.19_DATA_MANAGEMENT/Temp_TestMergin/Test_unique_constraints.gpkg',
'OUTPUT_LAYERS': ['//tsclient/P/LiveData/M1.1.19_DATA_MANAGEMENT/Temp_TestMergin/Test_unique_constraints.gpkg|layername=test_unique']}

Loading resulting layers
Algorithm 'Package layers' finished

### Versions

QGIS version
3.16.5-Hannover
QGIS code revision
58ba7c1ed6
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.1.4
Running against GDAL/OGR
3.1.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows Server 2016 (10.0)
Active python plugins
GroupStats; 
plaingeometryeditor; 
quick_map_services; 
TomBio; 
db_manager; 
MetaSearch; 
processing

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

I am using the version of QGIS that my organisation allows, we don't have the latest one because it was causing problems for our system, I don't know the exact details.

I have NOT been able to use a new QGIS profile, I did try but it does not have any user permissions so I can't access my organisations' database to do anything.

### PR DESCRIPTION
**Pull request by elpaso**
Fix #58431

I feel that the underlying issue remains (which is that setting the strength to NotSet removes it from the field but not from the layer and that the default strength is Hard) but since it was clearly made in purpose 8 years ago I am not keen to change it lightly.

### CODE REVIEW

**Comment by elpaso** [comment]
@nyalldawson this fixes the issue https://github.com/qgis/QGIS/issues/58341 but not the root cause and I'd like an opinion about the possibility to change
https://github.com/qgis/QGIS/blob/master/src/core/qgsfieldconstraints.cpp#L32
(see also: https://github.com/qgis/QGIS/blob/master/src/core/qgsfieldconstraints.cpp#L40) to be consistent with QgsVectorLayer::setFieldConstraint which doesn't remove the constraint if strength is NotSet.


The current behavior is not consistent.

**Comment by nyalldawson** [comment]
@elpaso 

> default strength is Hard

It's too long ago for me to remember exactly, but I suspect this was done in order to preserve API (ie before the introducing of soft constraints, all constraints where implicitly "hard" but we weren't explicitly tracking that). 

I'm happy for you to tweak this logic as needed -- there's good test coverage there, so I'm confident we'll be alerted if you break anything in the process..

File: src/core/qgsfieldconstraints.h [hunk 1]
```diff
@@ -89,7 +89,8 @@ class CORE_EXPORT QgsFieldConstraints
 
     /**
      * Returns the strength of a field constraint, or ConstraintStrengthNotSet if the constraint
-     * is not present on this field.
+     * is not present on this field. If the strength is not set returns ConstraintStrengthNotSet
+     * for anything but ConstraintExpression when returns ConstraintStrengthHard.
```
Comment (nyalldawson) [inline src/core/qgsfieldconstraints.h:93]:
```suggestion
     * for anything but ConstraintExpression which returns ConstraintStrengthHard.
```
?

Comment (elpaso) [reply src/core/qgsfieldconstraints.h]: done

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**